src/u2fprotocol.rs

```diff
@@ -322,135 +214,91 @@ fn sendrecv<T>(dev: &mut T, cmd: u8, send: &[u8]) -> io::Result<Vec<u8>>
 where
     T: U2FDevice + Read + Write,
 {
-    let mut sequence: u8 = 0; // Start at 0
-    let mut data_itr = send.into_iter();
-    let mut init_sent = false;
+    let mut sequence = 0u8;
+    let mut data = send;
+
     // Write Data.
-    while data_itr.size_hint().0 != 0 {
-        // Add 1 to HID_RPT_SIZE since we need to prefix this with a record
-        // index.
-        let mut frame: [u8; HID_RPT_SIZE + 1] = [0; HID_RPT_SIZE + 1];
-        if !init_sent {
-            let mut uf = U2FHIDInit {
-                cid: dev.get_cid(),
-                cmd: cmd,
-                bcnth: (send.len() >> 8) as u8,
-                bcntl: send.len() as u8,
-                data: [0; INIT_DATA_SIZE],
-            };
-            set_data(&mut uf.data, &mut data_itr, INIT_DATA_SIZE);
-            frame[1..].clone_from_slice(to_u8_array(&uf));
-            init_sent = true;
+    while data.len() > 0 {
+        // HID_RPT_SIZE+1 since we need to prefix this with a record index.
+        let mut frame = [0; HID_RPT_SIZE + 1];
+
+        if data.len() == send.len() {
+            let max = cmp::min(data.len(), INIT_DATA_SIZE);
+            let uf = U2FHIDInit::new(dev, cmd, send.len(), &data[..max]);
+            uf.to_bytes(&mut frame[1..]);
+            data = &data[max..];
         } else {
-            let mut uf = U2FHIDCont {
-                cid: dev.get_cid(),
-                seq: sequence,
-                data: [0; CONT_DATA_SIZE],
-            };
-            set_data(&mut uf.data, &mut data_itr, CONT_DATA_SIZE);
+            // First cont package has seq=1.
+            let max = cmp::min(data.len(), CONT_DATA_SIZE);
+            let uf = U2FHIDCont::new(dev, sequence, &data[..max]);
+            uf.to_bytes(&mut frame[1..]);
+            data = &data[max..];
             sequence += 1;
-            frame[1..].clone_from_slice(to_u8_array(&uf));
         }
 
         if log_enabled!(log::LogLevel::Trace) {
             let parts: Vec<String> = frame.iter().map(|byte| format!("{:02x}", byte)).collect();
             trace!("USB send: {}", parts.join(""));
         }
 
-        if let Err(er) = dev.write(&frame) {
-            return Err(er);
-        };
+        try!(dev.write(&frame));
```

Comment:
Good point. I got confused by outdated docs and used `try!` in other places. The docs say:

> Prefer using ? syntax to try!. ? is built in to the language and is more succinct than try!. It is the standard method for error propagation.

Let's use `?`.